hpx/runtime/serialization/access.hpp

```diff
@@ -36,6 +34,32 @@ namespace hpx { namespace serialization
 
     class access
     {
+        template <class T>
+        class has_serialize
+        {
+            struct two_chars
+            {
+                char _[2];
+            };
+
+            template <class T1> static char test(...);
+
+            // the following expression sfinae trick
+            // appears to work on clang-3.4, gcc-4.9,
+            // icc-16, msvc-2017 (at least)
+            // note that this detection would have been much easier
+            // to implement if there hadn't been an issue with gcc:
+            // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82478
+            template <class T1, class = decltype(
+                    std::declval<typename std::remove_const<T1>::type &>().
+                        serialize(std::declval<output_archive &>(), 0u))>
+            static two_chars test(int);
+
+        public:
+            static constexpr bool value =
+                sizeof(test<T>(0)) == sizeof(two_chars);
```

Comment:
related to the above discussion ... I was thinking of why for example `value = decltype(test<T>(0))::value;` didn't work. It looks like this is unrelated to the discussion in #3134. If this is the only way to get it to work, there's nothing one can do, right?